File: lib/utils/seed_scholarships.dart
import 'package:cloud_firestore/cloud_firestore.dart';
import 'package:flutter/foundation.dart';
import '../models/scholarship.dart';

Future<void> seedScholarships() async {
  final firestore = FirebaseFirestore.instance;
  final batch = firestore.batch();
  final collection = firestore.collection('scholarships');

  final List<Scholarship> seeds = [
    // 1. National Merit Scholarship
    Scholarship(
      id: collection.doc().id,
      title: 'National Merit Scholarship',
      organization: 'Min. of Education',
      description:
          'For students with exceptional academic performance (CGPA > 9.0). Open to all categories.',
      amount: 50000,
      eligibilityCriteria: EligibilityCriteria(
        minIncome: 0,
        maxIncome: 1000000, // High limit
        categories: ['General', 'OBC', 'SC', 'ST', 'EWS'],
        courses: ['B.Tech', 'M.Tech', 'B.Sc', 'M.Sc'],
        years: [1, 2, 3, 4],
      ),
      deadline: DateTime.now().add(const Duration(days: 30)),
      createdBy: 'system_seed',
      createdAt: DateTime.now(),
    ),
    // 2. Post-Matric Scholarship for SC Students
    Scholarship(
      id: collection.doc().id,
      title: 'Post-Matric Scholarship (SC)',
      organization: 'Min. of Social Justice',
      description:
          'Financial assistance for Scheduled Caste students pursuing higher education.',
      amount: 30000,
      eligibilityCriteria: EligibilityCriteria(
        minIncome: 0,
        maxIncome: 250000,
        categories: ['SC'],
        courses: ['B.Tech', 'M.Tech'],
        years: [1, 2, 3, 4],
      ),
      deadline: DateTime.now().add(const Duration(days: 45)),
      createdBy: 'system_seed',
      createdAt: DateTime.now(),
    ),
    // 3. Post-Matric Scholarship for ST Students
    Scholarship(
      id: collection.doc().id,
      title: 'Post-Matric Scholarship (ST)',
      organization: 'Min. of Tribal Affairs',
      description:
          'Financial assistance for Scheduled Tribe students pursuing higher education.',
      amount: 30000,
      eligibilityCriteria: EligibilityCriteria(
        minIncome: 0,
        maxIncome: 250000,
        categories: ['ST'],
        courses: ['B.Tech', 'M.Tech'],
        years: [1, 2, 3, 4],
      ),
      deadline: DateTime.now().add(const Duration(days: 45)),
      createdBy: 'system_seed',
      createdAt: DateTime.now(),
    ),
    // 4. EWS Support Grant
    Scholarship(
      id: collection.doc().id,
      title: 'Economically Weaker Section Grant',
      organization: 'Central Government',
      description:
          'Tuition support for students from Economically Weaker Sections with varied backgrounds.',
      amount: 25000,
      eligibilityCriteria: EligibilityCriteria(
        minIncome: 0,
        maxIncome: 100000,
        categories: ['EWS'],
        courses: ['Any'],
        years: [1, 2, 3, 4],
      ),
      deadline: DateTime.now().add(const Duration(days: 20)),
      createdBy: 'system_seed',
      createdAt: DateTime.now(),
    ),
    // 5. OBC Merit-cum-Means Scholarship
    Scholarship(
      id: collection.doc().id,
      title: 'OBC Merit-cum-Means',
      organization: 'National OBC Commission',
      description:
          'For OBC students demonstrating both merit and financial need.',
      amount: 20000,
      eligibilityCriteria: EligibilityCriteria(
        minIncome: 0,
        maxIncome: 300000,
        categories: ['OBC'],
        courses: ['B.Tech'],
        years: [2, 3, 4],
      ),
      deadline: DateTime.now().add(const Duration(days: 60)),
      createdBy: 'system_seed',
      createdAt: DateTime.now(),
    ),
    // 6. Girls in STEM Scholarship
    Scholarship(
      id: collection.doc().id,
      title: 'Women in STEM Initiative',
      organization: 'Dept. of Science & Tech',
      description:
          'Encouraging female students in Engineering and Science fields. Open to all categories.',
      amount: 40000,
      eligibilityCriteria: EligibilityCriteria(
        minIncome: 0,
        maxIncome: 800000,
        categories: [
          'General',
          'OBC',
          'SC',
          'ST',
          'EWS',
        ], // Logic handle gender separately or assume applied by girls
        courses: ['B.Tech', 'B.Sc'],
        years: [1],
      ),
      deadline: DateTime.now().add(const Duration(days: 90)),
      createdBy: 'system_seed',
      createdAt: DateTime.now(),
    ),
    // 7. Research Excellence Grant
    Scholarship(
      id: collection.doc().id,
      title: 'Research Excellence Grant',
      organization: 'University Grants Commission',
      description:
          'For postgraduate students engaged in advanced research projects.',
      amount: 60000,
      eligibilityCriteria: EligibilityCriteria(
        minIncome: 0,
        maxIncome: 1500000,
        categories: ['General', 'OBC', 'SC', 'ST', 'EWS'],
        courses: ['M.Tech', 'M.Sc', 'PhD'],
        years: [1, 2],
      ),
      deadline: DateTime.now().add(const Duration(days: 30)),
      createdBy: 'system_seed',
      createdAt: DateTime.now(),
    ),
    // 8. Minority Welfare Scholarship
    Scholarship(
      id: collection.doc().id,
      title: 'Minority Welfare Scholarship',
      organization: 'Min. of Minority Affairs',
      description: 'Support for students from minority communities.',
      amount: 15000,
      eligibilityCriteria: EligibilityCriteria(
        minIncome: 0,
        maxIncome: 200000,
        categories: [
          'General',
          'OBC',
        ], // Often mapped to specific religious minorities, simplifying here
        courses: ['Any'],
        years: [1, 2, 3, 4],
      ),
      deadline: DateTime.now().add(const Duration(days: 40)),
      createdBy: 'system_seed',
      createdAt: DateTime.now(),
    ),
    // 9. Sports Quota Scholarship
    Scholarship(
      id: collection.doc().id,
      title: 'Sports Achiever Scholarship',
      organization: 'Sports Authority of India',
      description:
          'For students who have represented the state/nation in sports.',
      amount: 25000,
      eligibilityCriteria: EligibilityCriteria(
        minIncome: 0,
        maxIncome: 5000000,
        categories: ['General', 'OBC', 'SC', 'ST', 'EWS'],
        courses: ['Any'],
        years: [1, 2, 3, 4],
      ),
      deadline: DateTime.now().add(const Duration(days: 15)),
      createdBy: 'system_seed',
      createdAt: DateTime.now(),
    ),
    // 10. Physically Challenged Student Aid
    Scholarship(
      id: collection.doc().id,
      title: 'Divyangjan Study Aid',
      organization: 'Dept. of Empowerment of PwDs',
      description:
          'Financial support for differently-abled students for study materials and aids.',
      amount: 35000,
      eligibilityCriteria: EligibilityCriteria(
        minIncome: 0,
        maxIncome: 600000,
        categories: ['General', 'OBC', 'SC', 'ST', 'EWS'],
        courses: ['Any'],
        years: [1, 2, 3, 4],
      ),
      deadline: DateTime.now().add(const Duration(days: 50)),
      createdBy: 'system_seed',
      createdAt: DateTime.now(),
    ),
  ];

  for (final scholarship in seeds) {
    batch.set(collection.doc(scholarship.id), scholarship.toFirestore());
  }

  await batch.commit();
  debugPrint('Seeded 10 scholarships successfully!');
}

File: scripts/python/seed_scholarships.py

import firebase_admin
from firebase_admin import credentials
from firebase_admin import firestore
import datetime
import os

# Configuration
SERVICE_ACCOUNT_KEY = '../../serviceAccountKey.json'

def initialize_firebase():
    if not firebase_admin._apps:
        cred = credentials.Certificate(SERVICE_ACCOUNT_KEY)
        firebase_admin.initialize_app(cred)
    return firestore.client()

def seed_scholarships():
    if not os.path.exists(SERVICE_ACCOUNT_KEY):
        print(f"❌ Error: Service account key not found: {SERVICE_ACCOUNT_KEY}")
        return

    db = initialize_firebase()
    
    # --- Delete Existing Scholarships first ---
    print("🗑️ Deleting existing scholarships...")
    docs = db.collection('scholarships').stream()
    deleted_count = 0
    batch = db.batch()
    
    for doc in docs:
        batch.delete(doc.reference)
        deleted_count += 1
        if deleted_count % 400 == 0: # Commit every 400 deletions to be safe within limit (500)
             batch.commit()
             batch = db.batch()
             
    batch.commit() # Commit remainder
    print(f"  ✅ Deleted {deleted_count} existing scholarships.")

    print("\n🌱 Seeding Scholarships...")
    
    scholarships = [
        # --- Original 5 (Updated Schema) ---
        {
            'name': 'Merit Excellence Scholarship',
            'provider': 'Techno India NJR',
            'amount': 50000,
            'description': 'Awarded to students with a CGPA above 9.0 in the previous semester.',
            'deadline': datetime.datetime.now() + datetime.timedelta(days=45),
            'category': 'Merit',
            'isActive': True,
            'createdAt': firestore.SERVER_TIMESTAMP,
            'eligibilityCriteria': {
                'minIncome': 0,
                'maxIncome': 99999999,
                'minCGPA': 9.0,
                'minAttendance': 0,
                'categories': [],
                'courses': [],
                'years': []
            }
        },
        {
            'name': 'Need-Based Financial Aid',
            'provider': 'Government of India',
            'amount': 25000,
            'description': 'Support for students from economically weaker sections. Valid income certificate required.',
            'deadline': datetime.datetime.now() + datetime.timedelta(days=30),
            'category': 'Government',
            'isActive': True,
            'createdAt': firestore.SERVER_TIMESTAMP,
            'eligibilityCriteria': {
                'minIncome': 0,
                'maxIncome': 500000,
                'minCGPA': 0,
                'minAttendance': 0,
                'categories': [],
                'courses': [],
                'years': []
            }
        },
        {
            'name': 'Women in Tech Grant',
            'provider': 'Google & Women Techmakers',
            'amount': 100000,
            'description': 'Encouraging female students pursuing degrees in Computer Science and Engineering.',
            'deadline': datetime.datetime.now() + datetime.timedelta(days=60),
            'category': 'Private',
            'isActive': True,
            'createdAt': firestore.SERVER_TIMESTAMP,
            'eligibilityCriteria': {
                'minIncome': 0,
                'maxIncome': 99999999,
                'minCGPA': 0,
                'minAttendance': 0,
                'categories': [],
                # Updated to use Full Branch Names
                'courses': ['Computer Engineering', 'Information Technology', 'AI & DS'], 
                'years': []
            }
        },
        {
            'name': 'Sports Achievement Award',
            'provider': 'State Sports Council',
            'amount': 15000,
            'description': 'For students who have represented the college at state or national level sports.',
            'deadline': datetime.datetime.now() + datetime.timedelta(days=15),
            'category': 'Sports',
            'isActive': True,
            'createdAt': firestore.SERVER_TIMESTAMP,
            'eligibilityCriteria': {
                'minIncome': 0,
                'maxIncome': 99999999,
                'minCGPA': 0,
                'minAttendance': 50.0,
                'categories': [],
                'courses': [],
                'years': []
            }
        },
        {
            'name': 'Alumni Association Grant',
            'provider': 'TINJR Alumni',
            'amount': 20000,
            'description': 'Funded by our alumni network to support students showing exceptional leadership.',
            'deadline': datetime.datetime.now() + datetime.timedelta(days=90),
            'category': 'Private',
            'isActive': True,
            'createdAt': firestore.SERVER_TIMESTAMP,
            'eligibilityCriteria': {
                'minIncome': 0,
                'maxIncome': 99999999,
                'minCGPA': 7.0,
                'minAttendance': 0,
                'categories': [],
                'courses': [],
                'years': []
            }
        },
        
        # --- New 10 ---
        {
            'name': 'Global Tech Innovators Grant',
            'provider': 'Tech Giants Consortium',
            'amount': 120000,
            'description': 'A prestigious grant for exceptional students in computer science and technology fields.',
            'deadline': datetime.datetime.now() + datetime.timedelta(days=45),
            'category': 'Private',
            'isActive': True,
            'createdAt': firestore.SERVER_TIMESTAMP,
            'eligibilityCriteria': {
                'minIncome': 0,
                'maxIncome': 99999999,
                'minCGPA': 8.5,
                'minAttendance': 0,
                'categories': [],
                # Updated to use Full Branch Names
                'courses': ['Computer Engineering', 'Information Technology', 'AI & DS', 'Electronics & Telecom'],
                'years': [3, 4]
            }
        },
        {
            'name': 'Future Leaders Scholarship',
            'provider': 'National Leadership Foundation',
            'amount': 40000,
            'description': 'Recognizing students with consistent academic performance and high attendance.',
            'deadline': datetime.datetime.now() + datetime.timedelta(days=30),
            'category': 'Merit',
            'isActive': True,
            'createdAt': firestore.SERVER_TIMESTAMP,
            'eligibilityCriteria': {
                'minIncome': 0,
                'maxIncome': 99999999,
                'minCGPA': 7.5,
                'minAttendance': 85.0,
                'categories': [],
                'courses': [],
                'years': []
            }
        },
        {
            'name': 'Community Support Fund',
            'provider': 'State Government',
            'amount': 25000,
            'description': 'Financial assistance for students from economically weaker sections.',
            'deadline': datetime.datetime.now() + datetime.timedelta(days=60),
            'category': 'Government',
            'isActive': True,
            'createdAt': firestore.SERVER_TIMESTAMP,
            'eligibilityCriteria': {
                'minIncome': 0,
                'maxIncome': 300000,
                'minCGPA': 0,
                'minAttendance': 60.0,
                'categories': [],
                'courses': [],
                'years': []
            }
        },
        {
            'name': 'Women in STEM Fellowship',
            'provider': 'Women Techmakers',
            'amount': 75000,
            'description': 'Empowering female students pursuing degrees in Science and Engineering.',
            'deadline': datetime.datetime.now() + datetime.timedelta(days=90),
            'category': 'Private',
            'isActive': True,
            'createdAt': firestore.SERVER_TIMESTAMP,
            'eligibilityCriteria': {
                'minIncome': 0,
                'maxIncome': 99999999,
                'minCGPA': 7.0,
                'minAttendance': 0,
                'categories': [],
                # Updated to use Full Branch Names
                'courses': ['Computer Engineering', 'Information Technology', 'Electronics & Telecom', 'Mechanical Engineering', 'Civil Engineering', 'Electrical Engineering'],
                'years': []
            }
        },
        {
            'name': 'State Sports Excellence Award',
            'provider': 'State Sports Council',
            'amount': 30000,
            'description': 'For students who have demonstrated excellence in sports activities.',
            'deadline': datetime.datetime.now() + datetime.timedelta(days=20),
            'category': 'Sports',
            'isActive': True,
            'createdAt': firestore.SERVER_TIMESTAMP,
            'eligibilityCriteria': {
                'minIncome': 0,
                'maxIncome': 99999999,
                'minCGPA': 5.0, 
                'minAttendance': 50.0,
                'categories': [],
                'courses': [],
                'years': []
            }
        },
        {
            'name': 'First Year Merit Scholarship',
            'provider': 'College Trust',
            'amount': 15000,
            'description': 'Welcoming bright minds. Exclusive for 1st-year students with outstanding entry scores.',
            'deadline': datetime.datetime.now() + datetime.timedelta(days=15),
            'category': 'Merit',
            'isActive': True,
            'createdAt': firestore.SERVER_TIMESTAMP,
            'eligibilityCriteria': {
                'minIncome': 0,
                'maxIncome': 99999999,
                'minCGPA': 8.0,
                'minAttendance': 0,
                'categories': [],
                'courses': [],
                'years': [1]
            }
        },
        {
            'name': 'Final Year Project Grant',
            'provider': 'Innovation Hub',
            'amount': 50000,
            'description': 'Funding for final year students to support their capstone projects and research.',
            'deadline': datetime.datetime.now() + datetime.timedelta(days=40),
            'category': 'Research',
            'isActive': True,
            'createdAt': firestore.SERVER_TIMESTAMP,
            'eligibilityCriteria': {
                'minIncome': 0,
                'maxIncome': 99999999,
                'minCGPA': 6.5,
                'minAttendance': 75.0,
                'categories': [],
                'courses': [],
                'years': [4]
            }
        },
        {
            'name': 'Minority Empowerment Scheme',
            'provider': 'Ministry of Minority Affairs',
            'amount': 35000,
            'description': 'Supporting students from minority communities (SC/ST/OBC).',
            'deadline': datetime.datetime.now() + datetime.timedelta(days=25),
            'category': 'Government',
            'isActive': True,
            'createdAt': firestore.SERVER_TIMESTAMP,
            'eligibilityCriteria': {
                'minIncome': 0,
                'maxIncome': 800000,
                'minCGPA': 6.0,
                'minAttendance': 70.0,
                'categories': ['SC', 'ST', 'OBC'],
                'courses': [],
                'years': []
            }
        },
        {
            'name': 'Dr. APJ Abdul Kalam Research Fellowship',
            'provider': 'National Science Foundation',
            'amount': 100000,
            'description': 'For students showing exceptional promise in scientific research and innovation.',
            'deadline': datetime.datetime.now() + datetime.timedelta(days=120),
            'category': 'Research',
            'isActive': True,
            'createdAt': firestore.SERVER_TIMESTAMP,
            'eligibilityCriteria': {
                'minIncome': 0,
                'maxIncome': 99999999,
                'minCGPA': 9.0,
                'minAttendance': 90.0,
                'categories': [],
                'courses': [],
                'years': [3, 4]
            }
        },
        {
            'name': 'Arts & Culture Grant',
            'provider': 'Heritage Foundation',
            'amount': 10000,
            'description': 'For students actively involved in cultural activities and arts.',
            'deadline': datetime.datetime.now() + datetime.timedelta(days=10),
            'category': 'Private',
            'isActive': True,
            'createdAt': firestore.SERVER_TIMESTAMP,
            'eligibilityCriteria': {
                'minIncome': 0,
                'maxIncome': 99999999,
                'minCGPA': 6.0,
                'minAttendance': 60.0,
                'categories': [],
                # Updated to use valid branches or left empty if generic
                'courses': [], 
                'years': []
            }
        },
    ]

    count = 0
    for sch in scholarships:
        db.collection('scholarships').add(sch)
        print(f"  ✅ Created scholarship: {sch['name']}")
        count += 1

    print(f"\n🎉 Successfully created {count} scholarships.")

if __name__ == "__main__":
    seed_scholarships()
